// slash-create/lib/servers/azure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureFunctionServer = void 0;
const server_1 = require("../server");
/**
 * A server for Azure Function integration
 * @see https://docs.microsoft.com/en-us/azure/azure-functions/
 */
class AzureFunctionServer extends server_1.Server {
    constructor(moduleExports, target = 'interactions') {
        super({ alreadyListening: true });
        moduleExports[target] = this._onRequest.bind(this);
    }
    _onRequest(context, req) {
        if (!this._handler) {
            context.res.status = 503;
            context.res.send('Server has no handler');
        }
        if (req.method !== 'POST') {
            context.res.status = 400;
            context.res.send('Server only supports POST requests.');
        }
        this._handler({
            headers: req.headers,
            body: req.body,
            request: req,
            response: context.res
        }, async (response) => {
            context.res.status = response ? response.status || 200 : 200;
            context.res.header('Content-type', 'application/json');
            if (response) {
                for (const key in response.headers)
                    context.res.header(key, response.headers[key]);
            }
            context.res.send(response.body);
        });
    }
    /** @private */
    createEndpoint(path, handler) {
        this._handler = handler;
    }
}
exports.AzureFunctionServer = AzureFunctionServer;
